core/src/main/scala/shapeless/generic.scala

```diff
@@ -306,8 +306,10 @@ trait CaseClassMacros extends ReprTypes {
       appliedType(tpe, List(typeOf[Any])).dealias
     else tpe
 
-  def isProductAux(tpe: Type): Boolean =
-    tpe.typeSymbol.isClass && (isCaseClassLike(classSym(tpe)) || HasApplyUnapply(tpe) || HasCtorUnapply(tpe))
+  def isProductAux(tpe: Type): Boolean = tpe.typeSymbol.isClass && {
+    val cls = classSym(tpe)
+    isCaseObjectLike(cls) || isCaseClassLike(cls) || HasApplyUnapply(tpe) || HasCtorUnapply(tpe)
```

Comment:
The `isCaseObjectLike` check is (mostly) redundant, but it's much cheaper than `isCaseClassLike` when the object is not `case`. It will also cover objects with fields which failed before:
```scala
object O { val x = 42 }
Generic[O.type]
```